### PR TITLE
test/asn1_time_test.c: Better check of signed time_t

### LINKED ISSUE #4177: ASN1_TIME failure on VMS
I noticed today that `[.test.recipes]90-test_asn1_time.t` currently fails on VMS, so I checked what happens by running the test executable directly, and well...

```
$ run [.test]asn1_time_test
# INFO: 
# Adding negative-sign time_t tests
1..2
    # Subtest: test_table_pos
    1..46
    ok 1 - iteration 1
    ok 2 - iteration 2
    ok 3 - iteration 3
    ok 4 - iteration 4
    ok 5 - iteration 5
    ok 6 - iteration 6
    ok 7 - iteration 7
    ok 8 - iteration 8
    ok 9 - iteration 9
    ok 10 - iteration 10
    ok 11 - iteration 11
    ok 12 - iteration 12
    ok 13 - iteration 13
    ok 14 - iteration 14
    ok 15 - iteration 15
    ok 16 - iteration 16
    ok 17 - iteration 17
    ok 18 - iteration 18
    ok 19 - iteration 19
    ok 20 - iteration 20
    ok 21 - iteration 21
    ok 22 - iteration 22
    ok 23 - iteration 23
    ok 24 - iteration 24
    ok 25 - iteration 25
    ok 26 - iteration 26
    ok 27 - iteration 27
    ok 28 - iteration 28
    ok 29 - iteration 29
    ok 30 - iteration 30
    ok 31 - iteration 31
    ok 32 - iteration 32
    ok 33 - iteration 33
    ok 34 - iteration 34
    ok 35 - iteration 35
    ok 36 - iteration 36
    ok 37 - iteration 37
    ok 38 - iteration 38
    ok 39 - iteration 39
    ok 40 - iteration 40
    ok 41 - iteration 41
    ok 42 - iteration 42
    ok 43 - iteration 43
    ok 44 - iteration 44
    ok 45 - iteration 45
    ok 46 - iteration 46
ok 1 - test_table_pos
    # Subtest: test_table_neg
    1..3
    # ERROR: (int) 'ASN1_TIME_cmp_time_t(&atime, td->t) == 0' failed @ PROGRAM:[LEVITTE-BUILDS.DOWNLOADS.openssl-SNAP-20170816.test]asn1_time_test.c;1:129
    # [-1] compared to [0]
    # INFO: 
    # ASN1_TIME_cmp_time_t(19011213204552Z vs -2147483648) compare failed
    # ERROR: (int) 'ptime->type == td->expected_type' failed @ PROGRAM:[LEVITTE-BUILDS.DOWNLOADS.openssl-SNAP-20170816.test]asn1_time_test.c;1:169
    # [23] compared to [24]
    # INFO: 
    # ASN1_TIME_set(-2147483648) unexpected type
    # INFO: 
    # ASN1_TIME_set() = 380119031408Z
    # ERROR: 
    # atime=19011213204552Z
    # 
    not ok 1 - iteration 1
    # ERROR: (int) 'ASN1_TIME_cmp_time_t(&atime, td->t) == 0' failed @ PROGRAM:[LEVITTE-BUILDS.DOWNLOADS.openssl-SNAP-20170816.test]asn1_time_test.c;1:129
    # [-1] compared to [0]
    # INFO: 
    # ASN1_TIME_cmp_time_t(691006121456Z vs -7472704) compare failed
    # ERROR: (int) 'ptime->type == td->expected_type' failed @ PROGRAM:[LEVITTE-BUILDS.DOWNLOADS.openssl-SNAP-20170816.test]asn1_time_test.c;1:169
    # [24] compared to [23]
    # INFO: 
    # ASN1_TIME_set(-7472704) unexpected type
    # INFO: 
    # ASN1_TIME_set() = 21051112184312Z
    # ERROR: (int) 'ASN1_TIME_cmp_time_t(ptime, td->t) == 0' failed @ PROGRAM:[LEVITTE-BUILDS.DOWNLOADS.openssl-SNAP-20170816.test]asn1_time_test.c;1:243
    # [-1] compared to [0]
    # INFO: 
    # ASN1_TIME_to_generalizedtime(691006121456Z->19691006121456Z) bad result
    # ERROR: 
    # atime=691006121456Z
    # 
    not ok 2 - iteration 2
    # ERROR: (int) 'ASN1_TIME_cmp_time_t(&atime, td->t) == 0' failed @ PROGRAM:[LEVITTE-BUILDS.DOWNLOADS.openssl-SNAP-20170816.test]asn1_time_test.c;1:129
    # [-1] compared to [0]
    # INFO: 
    # ASN1_TIME_cmp_time_t(19691006121456Z vs -7472704) compare failed
    # ERROR: (int) 'ptime->type == td->expected_type' failed @ PROGRAM:[LEVITTE-BUILDS.DOWNLOADS.openssl-SNAP-20170816.test]asn1_time_test.c;1:169
    # [24] compared to [23]
    # INFO: 
    # ASN1_TIME_set(-7472704) unexpected type
    # INFO: 
    # ASN1_TIME_set() = 21051112184312Z
    # ERROR: 
    # atime=19691006121456Z
    # 
    not ok 3 - iteration 3
not ok 2 - test_table_neg
$ 
```

### PR DESCRIPTION
**Pull request by levitte**
Fixes #4177

### CODE REVIEW

**Comment by levitte** [comment]
I wonder why anyone imagined that `localtime` on -1 would necessarely return NULL to indicate that `time_t` is unsigned...

**Review by richsalz — APPROVED** [review]

**Comment by richsalz** [comment]
ya know, the ```!(t>0)``` really jangles, but okay if you don't change it.

**Comment by levitte** [comment]
> ya know, the ```!(t>0)``` really jangles, but okay if you don't change it.

Sure, but think of what a picky compiler will say if `time_t` is unsigned and we have `t < 0`...  VMS C has been known to be picky ;-)

**Comment by kaduk** [comment]
Well, ```t<=0``` at least can be true sometimes, but still gets warnings from picky compilers, IIRC.

**Comment by richsalz** [comment]
That's an excellent point; please add a comment to the first time it's used.

**Comment by levitte** [comment]
Comment added

**Review by richsalz — APPROVED** [review]
thanks!

**Comment by levitte** [comment]
I'll merge as soon as all the CIs clear

**Comment by levitte** [comment]
Ah, it seems the `ptm` check is still needed as well.  I've added a commit to that effect, with commentary.  Needs re-confirmation.

**Comment by richsalz** [comment]
reonfirm

File: test/asn1_time_test.c [hunk 1]
```diff
@@ -276,18 +276,31 @@ static int test_table_neg_64bit(int idx)
...
 {
+    /*
+     * On platforms where |time_t| is an unsigned integer, t will be a
+     * positive number.
+     *
+     * We check if we're on a platform with a signed |time_t| with '!(t > 0)'
+     * because some compilers are picky if you do 't < 0', or even 't <= 0'
+     * if |t| is unsigned.
+     */
     time_t t = -1;
```
Comment (paulidale) [inline test/asn1_time_test.c]: Would this benefit from a cast?
I.e.
```
time_t t = (time_t) -1;
```

Comment (paulidale) [inline test/asn1_time_test.c]: This isn't meant to stop a merge.

Comment (levitte) [inline test/asn1_time_test.c]: I haven't seen a case where that's needed.

**Comment by levitte** [comment]
Merged. fdf9450f4266971035ca5f87c46c3826b49b5757

**Pull request closed without merging**